runtime/src/iree/hal/drivers/hip/event_semaphore.c

```diff
@@ -99,484 +132,749 @@ static void iree_hal_hip_semaphore_destroy(
+  if (!iree_status_is_ok(status)) {
+    iree_slim_mutex_unlock(&semaphore->mutex);
+    IREE_TRACE_ZONE_END(z0);
+    return status;
+  }
+  iree_hal_hip_semaphore_work_item_t* work_item = NULL;
+  iree_hal_hip_semaphore_work_item_t* last_work_item = NULL;
+  if (value > semaphore->max_value_to_be_signaled) {
+    iree_hal_hip_util_tree_node_t* node = iree_hal_hip_util_tree_upper_bound(
+        &semaphore->event_queue.tree, semaphore->max_value_to_be_signaled);
+    // Collect all of the things to schedule now that we know we can safely make
+    // it to a given value.
+    while (node && iree_hal_hip_util_tree_node_get_key(node) <= value) {
+      iree_hal_hip_semaphore_queue_item_t* queue_item =
+          (iree_hal_hip_semaphore_queue_item_t*)
+              iree_hal_hip_util_tree_node_get_value(node);
+      iree_hal_hip_semaphore_work_item_t* next_work_item =
+          queue_item->work_item;
+      while (next_work_item) {
+        if (!work_item) {
+          work_item = next_work_item;
+        }
+        if (last_work_item && !last_work_item->next) {
+          last_work_item->next = next_work_item;
+        }
+        last_work_item = next_work_item;
+        next_work_item = next_work_item->next;
+      }
+      queue_item->work_item = NULL;
+      iree_hal_hip_util_tree_node_t* last_node = node;
+      node = iree_hal_hip_util_tree_node_next(node);
+      if (!queue_item->event) {
+        iree_hal_hip_util_tree_erase(&semaphore->event_queue.tree, last_node);
+      }
     }
   }
-  iree_slim_mutex_unlock(&semaphore->base.timepoint_mutex);
 
+  semaphore->max_value_to_be_signaled =
+      iree_max(semaphore->max_value_to_be_signaled, value);
+  iree_slim_mutex_unlock(&semaphore->mutex);
+
+  // Now that we have accumulated all of the work items, and we have
+  // unlocked the semaphore, start running through the work items.
+  while (work_item) {
+    iree_hal_hip_semaphore_work_item_t* next_work_item = work_item->next;
+    work_item->scheduled_callback(work_item->user_data, base_semaphore, status);
+    iree_allocator_free(semaphore->host_allocator, work_item);
+    work_item = next_work_item;
+  }
   IREE_TRACE_ZONE_END(z0);
-  return *out_event != NULL;
+  return status;
 }
 
-// Checks if the semaphore has to wait to reach `value`.
-// If it has to wait, then acquires a wait timepoint and returns it.
-// If we don't need to wait, then *out_timepoint is set to NULL.
-static iree_status_t iree_hal_hip_semaphore_try_wait_or_acquire_wait_timepoint(
+iree_status_t iree_hal_hip_semaphore_get_hip_event(
     iree_hal_semaphore_t* base_semaphore, uint64_t value,
-    iree_timeout_t timeout, iree_hal_hip_timepoint_t** out_timepoint) {
-  *out_timepoint = NULL;
+    iree_hal_hip_event_pool_t* event_pool,
+    iree_hal_hip_event_t** out_hip_event) {
   iree_hal_hip_semaphore_t* semaphore =
       iree_hal_hip_semaphore_cast(base_semaphore);
+  *out_hip_event = NULL;
   IREE_TRACE_ZONE_BEGIN(z0);
```

Comment:
you'll want to check your traces and make sure these functions are worth tracing - not all functions are, particularly queries that may be made many number of times - basically, if you end up with zones that are nanoseconds you're probably tracing something that makes more sense in aggregate (record the number of checks a semaphore did as a value instead of having a tree of query functions for each check that almost always return after one if statement)

it's a signal/noise ratio balancing kind of activity - traces aren't meant to replace debugging but instead show high-level program flow with focused areas that reveal tricky things - not all functions need to be traced. for small ones it's best to rely on sampling to see if they rise to the top (if you can't immediately tell the cost when authoring - it's always best to not do a lot of work rather than do a lot of work and trace it :)